 # Capstone Project Notebook

## Importing important libraries

In [33]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt



## Downloading the Loan data

In [34]:
!wget -O loan_train.csv https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/loan_train.csv

--2019-04-02 16:38:36--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/loan_train.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.193
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.193|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23101 (23K) [text/csv]
Saving to: ‘loan_train.csv’

100%[======================================>] 23,101      --.-K/s   in 0.002s  

2019-04-02 16:38:36 (13.3 MB/s) - ‘loan_train.csv’ saved [23101/23101]



In [35]:
# loading data from csv files
data = pd.read_csv('loan_train.csv')
data.head()

Unnamed: 0  Unnamed: 0.1 loan_status  Principal  terms effective_date  \
0           0             0     PAIDOFF       1000     30       9/8/2016   
1           2             2     PAIDOFF       1000     30       9/8/2016   
2           3             3     PAIDOFF       1000     15       9/8/2016   
3           4             4     PAIDOFF       1000     30       9/9/2016   
4           6             6     PAIDOFF       1000     30       9/9/2016   

    due_date  age             education  Gender  
0  10/7/2016   45  High School or Below    male  
1  10/7/2016   33              Bechalor  female  
2  9/22/2016   27               college    male  
3  10/8/2016   28               college  female  
4  10/8/2016   29               college    male

In [36]:
data.drop(data.iloc[:, 0:2], inplace = True, axis = 1) 
data.head()

loan_status  Principal  terms effective_date   due_date  age  \
0     PAIDOFF       1000     30       9/8/2016  10/7/2016   45   
1     PAIDOFF       1000     30       9/8/2016  10/7/2016   33   
2     PAIDOFF       1000     15       9/8/2016  9/22/2016   27   
3     PAIDOFF       1000     30       9/9/2016  10/8/2016   28   
4     PAIDOFF       1000     30       9/9/2016  10/8/2016   29   

              education  Gender  
0  High School or Below    male  
1              Bechalor  female  
2               college    male  
3               college  female  
4               college    male

In [37]:
print(data['loan_status'].value_counts())
print(data['education'].value_counts())
print(data['Gender'].value_counts())
#print(data['effective_date'].value_counts())
#print(data['due_date'].value_counts())

PAIDOFF       260
COLLECTION     86
Name: loan_status, dtype: int64
High School or Below    151
college                 149
Bechalor                 44
Master or Above           2
Name: education, dtype: int64
male      294
female     52
Name: Gender, dtype: int64


In [38]:
data.columns

Index(['loan_status', 'Principal', 'terms', 'effective_date', 'due_date',
       'age', 'education', 'Gender'],
      dtype='object')

In [39]:
X = data[['Principal', 'terms' , 'age', 'education', 'Gender']] .values  #.astype(float)
X[0:5]

array([[1000, 30, 45, 'High School or Below', 'male'],
       [1000, 30, 33, 'Bechalor', 'female'],
       [1000, 15, 27, 'college', 'male'],
       [1000, 30, 28, 'college', 'female'],
       [1000, 30, 29, 'college', 'male']], dtype=object)

In [40]:
y = data['loan_status'].values
y[0:5]

array(['PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF'], dtype=object)

In [41]:
from sklearn import preprocessing

le_gender = preprocessing.LabelEncoder()
le_gender.fit([ 'male', 'female'])
X[:,4] = le_gender.transform(X[:,4])
X[:,4]

le_education = preprocessing.LabelEncoder()
le_education.fit([ 'High School or Below', 'college','Bechalor','Master or Above'])
X[:,3] = le_education.transform(X[:,3])
X[:,3]



array([1, 0, 3, 3, 3, 3, 3, 3, 3, 1, 3, 3, 1, 3, 1, 1, 1, 3, 3, 1, 3, 1, 0,
       0, 3, 3, 1, 3, 3, 3, 3, 1, 0, 3, 3, 0, 3, 0, 1, 1, 3, 1, 0, 0, 1, 1,
       3, 1, 0, 1, 2, 0, 1, 3, 3, 0, 1, 0, 3, 1, 0, 3, 3, 1, 3, 1, 1, 3, 3,
       3, 1, 3, 1, 3, 1, 1, 1, 3, 3, 1, 0, 3, 3, 1, 3, 1, 1, 3, 3, 3, 3, 3,
       3, 3, 1, 3, 1, 0, 3, 3, 1, 3, 3, 3, 3, 1, 1, 0, 0, 1, 1, 1, 3, 3, 1,
       1, 3, 1, 1, 1, 3, 3, 1, 3, 3, 0, 0, 3, 3, 3, 3, 1, 1, 3, 1, 1, 0, 1,
       3, 1, 1, 3, 1, 1, 3, 1, 3, 1, 3, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3,
       3, 1, 1, 1, 3, 3, 3, 1, 3, 1, 3, 3, 0, 3, 3, 1, 1, 1, 1, 1, 1, 1, 3,
       1, 1, 1, 1, 3, 3, 1, 0, 1, 1, 0, 1, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 1,
       0, 1, 3, 1, 3, 1, 1, 0, 3, 1, 3, 3, 3, 3, 3, 1, 3, 1, 3, 1, 0, 0, 1,
       1, 3, 1, 3, 3, 1, 0, 3, 1, 3, 3, 3, 1, 0, 3, 1, 0, 0, 3, 0, 1, 3, 1,
       3, 1, 0, 3, 1, 1, 3, 3, 1, 3, 0, 3, 1, 3, 1, 3, 1, 1, 0, 0, 3, 1, 3,
       3, 3, 0, 0, 3, 0, 3, 1, 1, 3, 1, 3, 1, 3, 1, 1, 3, 2, 1, 0, 1, 1, 3,
       1, 3,

In [42]:
X[1:10]

array([[1000, 30, 33, 0, 0],
       [1000, 15, 27, 3, 1],
       [1000, 30, 28, 3, 0],
       [1000, 30, 29, 3, 1],
       [1000, 30, 36, 3, 1],
       [1000, 30, 28, 3, 1],
       [800, 15, 26, 3, 1],
       [300, 7, 29, 3, 1],
       [1000, 15, 39, 1, 1]], dtype=object)

In [43]:
X = preprocessing.StandardScaler().fit(X).transform(X.astype(float))
X[0:5]

/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/sklearn/utils/validation.py:475: DataConversionWarning: Data with input dtype object was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


array([[ 0.51578458,  0.92071769,  2.33152555, -0.64691859,  0.42056004],
       [ 0.51578458,  0.92071769,  0.34170148, -1.5212695 , -2.37778177],
       [ 0.51578458, -0.95911111, -0.65321055,  1.10178323,  0.42056004],
       [ 0.51578458,  0.92071769, -0.48739188,  1.10178323, -2.37778177],
       [ 0.51578458,  0.92071769, -0.3215732 ,  1.10178323,  0.42056004]])

In [44]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (276, 5) (276,)
Test set: (70, 5) (70,)


# KNN

In [45]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

In [46]:
classifier = KNeighborsClassifier(n_neighbors=5)  
classifier.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [62]:
# Best k
Ks=20
mean_acc=np.zeros((Ks-1))
std_acc=np.zeros((Ks-1))
for n in range(1,Ks):
    #Train Model and Predict  
    model = KNeighborsClassifier(n_neighbors=n).fit(X_train,y_train)
    yhat = model.predict(X_test)
    mean_acc[n-1]=np.mean(yhat==y_test);
mean_acc

array([ 0.65714286,  0.55714286,  0.68571429,  0.65714286,  0.77142857,
        0.71428571,  0.74285714,  0.72857143,  0.74285714,  0.71428571,
        0.77142857,  0.74285714,  0.8       ,  0.8       ,  0.8       ,
        0.8       ,  0.78571429,  0.78571429,  0.78571429])

In [63]:
# k=13 is the optimal value of k
k=13
classifier = KNeighborsClassifier(n_neighbors=k)  
classifier.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=13, p=2,
           weights='uniform')

In [64]:
yhat = classifier.predict(X_test)
yhat[0:5]

array(['PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF'], dtype=object)

In [65]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
print("Test set jaccard_similarity_score: ", jaccard_similarity_score(y_test, yhat, normalize=True))
print("Test set F1_score: ", f1_score(y_test, yhat, average='weighted'))


Test set jaccard_similarity_score:  0.8
Test set F1_score:  0.723790322581


## Decision Tree Classifier

In [50]:
from sklearn.tree import DecisionTreeClassifier

In [51]:
LoanTree = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
LoanTree # it shows the default parameters

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=4,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [52]:
LoanTree.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=4,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [53]:
yhat = LoanTree.predict(X_test)

In [54]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
print("Test set jaccard_similarity_score: ", jaccard_similarity_score(y_test, yhat, normalize=True))
print("Test set F1_score: ", f1_score(y_test, yhat, average='weighted'))


Test set jaccard_similarity_score:  0.785714285714
Test set F1_score:  0.691428571429


/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


# Support Vector Machine

In [55]:
from sklearn import svm
clf = svm.SVC(kernel='rbf')
clf.fit(X_train, y_train) 

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [56]:
yhat = clf.predict(X_test)
yhat [0:5]

array(['PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF'], dtype=object)

In [57]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
print("Test set jaccard_similarity_score: ", jaccard_similarity_score(y_test, yhat, normalize=True))
print("Test set F1_score: ", f1_score(y_test, yhat, average='weighted'))


Test set jaccard_similarity_score:  0.785714285714
Test set F1_score:  0.691428571429


/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


## Logistic Regression

In [58]:
le_status = preprocessing.LabelEncoder()
le_status.fit(['PAIDOFF','COLLECTION'])
y[:] = le_status.transform(y[:]).astype(float)

In [59]:
from sklearn.linear_model import LogisticRegression
LR = LogisticRegression(C=0.01, solver='liblinear').fit(X_train,y_train)
LR

LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [60]:
yhat = LR.predict(X_test)
yhat_prob = LR.predict_proba(X_test)
#yhat_prob

In [61]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
print("Test set jaccard_similarity_score: ", jaccard_similarity_score(y_test, yhat, normalize=True))
print("Test set F1_score: ", f1_score(y_test, yhat, average='weighted'))
print("Test set Log_loss: ", log_loss(y_test, yhat_prob))



Test set jaccard_similarity_score:  0.785714285714
Test set F1_score:  0.691428571429
Test set Log_loss:  0.60078003883


/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


# Final Report

| Algorithm          | Jaccard | F1-score | LogLoss |
|--------------------|---------|----------|---------|
| KNN                | 0.8     | 0.72     | NA      |
| Decision Tree      | 0.78    | 0.69     | NA      |
| SVM                | 0.78    | 0.69     | NA      |
| LogisticRegression | 0.78    | 0.69     | 0.60    |